circuit_knitting/cutting/cut_finding/best_first_search.py

```diff
@@ -0,0 +1,387 @@
+from .search_space_generator import SearchFunctions
+
+if TYPE_CHECKING:  # pragma: no cover
+    from .cut_optimization import CutOptimizationFuncArgs
+
+
+class BestFirstPriorityQueue:
+    """Class that implements priority queues for best-first search.
+
+    The tuples that are pushed onto the priority queues have the form:
+
+    (<cost>, <neg_search_depth>, <random_num>, <seq_count>, <search_state>),
+
+    where:
+
+    <cost> (numeric or tuple) is a numeric cost or tuple of numeric
+    lexically-ordered costs that are to be minimized.
+
+    <neg_search_depth> (int) is the negative of the search depth of the
+    search state represented by the tuple. Thus, if several search states
+    have identical costs, priority is given to the deepest states to
+    encourage depth-first behavior.
+
+    <random_num> is a pseudo-random number that randomly break ties in a
+    stable manner if several search states have identical costs at identical
+    search depths.
+
+    <seq_count> is a sequential count of push operations that is used to
+    break further ties just in case two states with the same costs and
+    depths are somehow assigned the same pseudo-random numbers.
+
+    <search_state> is a state object generated by the optimization process.
+    Because of the design of the tuple entries that precede it, state objects
+    never get evaluated in the heap-managment comparisons that are performed
+    internally by the priority-queue implementation.
+    """
+
+    def __init__(self, seed: int | None):
+        """Assign member variables."""
+        self.random_gen: Generator = np.random.default_rng(seed)
+        self.unique: count[int] = count()
+        self.pqueue: list[tuple] = list()
+
+    def put(
+        self,
+        state: DisjointSubcircuitsState,
+        depth: int,
+        cost: float | tuple[float, float],
+    ) -> None:
+        """Push state onto the priority queue.
+
+        The search depth and cost of the state must also be provided as input.
+        """
+        heapq.heappush(
+            self.pqueue,
+            (cost, (-depth), self.random_gen.random(), next(self.unique), state),
+        )
+
+    def get(
+        self,
+    ) -> tuple:
+        """Return the lowest cost state currently on the queue, along with its depth and cost.
+
+        None, None, None is returned if the priority queue is empty.
+        """
+        if self.qsize() == 0:  # pragma: no cover
+            return None, None, None
+
+        best: tuple = heapq.heappop(self.pqueue)
+
+        return best[-1], (-best[1]), best[0]
+
+    def qsize(self) -> int:
+        """Return the size of the priority queue."""
+        return len(self.pqueue)
+
+    def clear(self) -> None:
+        """Clear all entries in the priority queue."""
+        self.pqueue.clear()
+
+
+class BestFirstSearch:
+    """Implement Dijkstra's best-first search algorithm.
+
+     The search proceeds by choosing the deepest, lowest-cost state
```

Comment:
This entire docstring is indented 5 spaces instead of 4.  Might be nice to clean up before merging, but not a big deal.

Reply:
'Tis done: [3558bb7](https://github.com/Qiskit-Extensions/circuit-knitting-toolbox/pull/520/commits/3558bb733fd7a465b251a5dd918079b0be77f44b)